### ISSUE TITLE
How to check if the current invocation is running `publish`?

### ISSUE THREAD

**Issue by agocke** [dotnet/sdk#26324]
Title basically says it all -- but in certain cases I'd like to know in the project file whether we're doing a build or a publish. In particular, I might want a non-self-contained build by default, but always want a self-contained publish.

**Comment by dsplaisted** [comment]
There isn't a way to tell what targets MSBuild will run (before it actually runs them), so there's not a good way to do this.

On the other hand, we would like to be able to have different defaults and behavior between build and publish.  The first step toward that just went in with https://github.com/dotnet/sdk/pull/25991.

With that change, you could check the `_IsPublishing` property to see if the build was run via `dotnet publish`.  However, the property would not be set if you ran `dotnet build /t:Publish` or `msbuild /t:Publish`.

**Comment by marcpopMSFT** [comment]
@dsplaisted are there plans (or even capability) to set ispublishing when using the /t flag? Not sure what this issue tracks with 25991 merged now.

**Comment by dsplaisted** [comment]
No concrete plans I think.  We could detect and handle something like `/t:Publish` in the CLI, but that wouldn't handle cases like `/t:TargetThatDependsOnPublish`.  FYI @baronfel

**Comment by baronfel** [comment]
yeah, this is a sort of general problem we're hitting with all of these 'use publish to push better packaging practices'-style features - the gulf between `dotnet X` and `dotnet msbuild /t:X` grows over time. As I understand it @rainersigwald has been reluctant in the past (for good reason!) to add something like `@(MSBuildDesiredTarget)` as a known property, so we don't really have a good way to get out of this hole in a general fashion.

**Comment by dsplaisted** [comment]
> Aside -- could we just set `IsPublishing` in the Publish target? Is there some reason I'm missing why this isn't sufficient?

In that case it wouldn't be set during MSBuild evaluation or in targets that run before publish.  So I think in most cases it would be set way later than you would need to check it.

**Comment by agocke** [comment]
I see, in that case I could also set it to true in the Publish target in case of `/t:Publish`?

**Comment by nagilson** [comment]
> I see, in that case I could also set it to true in the Publish target in case of `/t:Publish`?

No, see this discussion here, it needs to be a property. https://github.com/dotnet/sdk/pull/25991#issuecomment-1164824894

**Comment by marcpopMSFT** [comment]
One suggestion is to capture anything specified through -t: on the commandline into a global property that's a semicolon separated list thing `dotnet build -t:Publish -t:Test` sets `$(UserTargets)=Publish;Tests`.  

Noah took care of `dotnet publish` and `dotnet pack` setting a global property of [_IsPublishing](https://github.com/dotnet/sdk/pull/25991/files#diff-bdc61f2233a5b57b4e92bd5364280e19b506f48eb7496fc2ba4286db10b3e95f)

**Comment by xoofx** [comment]
> yeah, this is a sort of general problem we're hitting with all of these 'use publish to push better packaging practices'-style features - the gulf between `dotnet X` and `dotnet msbuild /t:X` grows over time. As I understand it @rainersigwald has been reluctant in the past (for good reason!) to add something like `@(MSBuildDesiredTarget)` as a known property, so we don't really have a good way to get out of this hole in a general fashion.

Hey, I stumbled on this issue, as I wanted to set `PublishAot` by default to `true` in a target file, but only if we are in publish mode, so now I'm using `_IsPublishing` and it is okish.

But I'm wondering why it would not be desirable to have `MSBuildDesiredTarget` (or a property `$MSBuildDesiredTarget_$(ActualTarget)` that would e.g set `$MSBuildDesiredTarget_Build` or `$MSBuildDesiredTarget_Publish` to `true` if they are targeted)? It does seem that we have some usecases here that are indicating that it would be more convenient to have it built-in than to hack around it specifically for e.g Publish?

**Comment by xoofx** [comment]
Thanks for the explanation @nagilson, I appreciate the details!

> It is 'impossible' to know for sure which target is going to run early enough for MSBuild itself to take action. One reason is that people can implement their own custom targets that call publish or build later, so from the beginning of evaluation we cannot check the MSBuild target stack to see if Publish is in the target-stack.

Indeed, I understand that it's safer to plug it in the dotnet command. If I recall, it's not the only property that some commands are already setting or that the interaction with MSBuild is tweaked from there (e.g build => restore + build in separate steps)

Hope that `_IsPublishing` will stay stable, always a bit worried when I see a `_` in front of a property, as I have always assumed that they are considered as "`private`" MSBuild properties in the past that we should not rely on! 😅 

> BTW, ❤️ Unity

Same, will love it even more once we have migrated to .NET 7/8 + MSBuild! 😄

**Comment by MichalStrehovsky** [comment]
Did we consider doing a breaking change and erroring out in the `Publish` target if `_IsPublishing` is not set?

The problem is that we're getting more and more build customizations conditioned under `_IsPublishing`. External packages have similar needs (most recently https://github.com/microsoft/CsWinRT/pull/1547). We're now ending up in a situation where `dotnet build /t:Publish` will do very different things from `dotnet publish`. It might not be obvious what things are different, why, and how to fix it. Requiring to set a property before one is allowed to run the target could be one way to address that. This assumes we'd also make the property public so that people who want to run publish target can set it using a supported mechanism.

**Comment by nagilson** [comment]
We never considered that. I can see the logic behind it, though it would break a large number of customers and probably cause frustration. That being said, it is an interesting idea to solve this issue that you mention regarding _IsPublishing.

**Comment by MichalStrehovsky** [comment]
https://github.com/search?q=_IsPublishing+language%3AXML&type=code&l=XML - looks like many third parties depend on `_IsPublishing`. All of those uses will not do the intended thing when someone runs `/t:Publish` instead of the real publish.

I guess the question is at what point will `/t:Publish` without `/p:_IsPublishing=true` become more breaking than just erroring out with a prescriptive/self-serviceable message.

We're starting to get blocked internally on things because `/t:Publish` without `/p:_IsPublishing=true` is a thing: https://github.com/dotnet/sdk/pull/37872#issuecomment-1935428434

**Comment by nagilson** [comment]
Wow. `_IsPublishing` has only been around for about a year. As the creator of this thing, I can tell you we didn't expect that given the _.

Anyway, I think your point is interesting. Since @rainersigwald owns MS Build I think he would have to have a say on what we'd do here, as well as our PMs. But its something we could consider.

**Comment by nagilson** [comment]
Definitely not a 'good first issue'

**Comment by agocke** [comment]
also + @dsplaisted

**Comment by nagilson** [comment]
The above approach would also likely require changes to visual studio with @vijayrkn. I am curious about where @MichalStrehovsky's idea would go, though that would break a lot of people and cause every single person using t:Publish to fail until they update their properties to add `IsPublishing`, lest MSBuild made that change, which is a very difficult change to make per @rainersigwald as it would not support publish targets being injected later on in the build.

An alternative is to make Publish a Configuration or to rewrite the Publish properties into Release properties. Those are both very large undertakings.

**Comment by am11** [comment]
There is a similar convention in SDK which indicates the SDK action that can be performed on the project: [`<IsPublishable>`](https://learn.microsoft.com/dotnet/core/project-sdk/msbuild-props#ispublishable) and [`<IsPackable>`](https://learn.microsoft.com/nuget/reference/msbuild-targets#pack-target-inputs). Having similar `<IsPublishing>` and `<IsPacking>` when those actions are effective (set at the beginning of their corresponding targets) sounds reasonable.
